src/arc_spice/eval/inference_utils.py

```diff
@@ -0,0 +1,133 @@
+
+    def get_results(
+        self,
+        clean_output: dict[str, dict],
+        var_output: dict[str, dict],
+        test_row: dict[str, list[int]],
+        results_dict: dict[str, dict[str, list]],
+    ):
+        for step_name in clean_output:
+            results_dict = self.func_map[step_name](
+                test_row, clean_output, var_output, results_dict
+            )
+        results_dict["input_data"]["celex_ids"].append(test_row["celex_id"])
+        return results_dict
+
+    def recognition_results(self, test_row, clean_output, var_output, results_dict):
+        assert test_row is not None
+        assert clean_output is not None
+        assert var_output is not None
+        # ### RECOGNITION ###
+        # TODO: add this into results_getter
+        return results_dict
+
+    def translation_results(self, test_row, clean_output, var_output, results_dict):
+        # ### TRANSLATION ###
+        source_text = test_row["target_text"]
+        target_text = test_row["target_text"]
+        clean_translation = clean_output["translation"]["full_output"]
+
+        # load error model
+        comet_inp = [
+            {
+                "src": source_text,
+                "mt": clean_translation,
+                "ref": target_text,
+            }
+        ]
+        # comet doesn't work on MPS
+        comet_device = "cuda" if torch.cuda.is_available() else "cpu"
+        comet_output = self.comet_model.predict(
+            comet_inp, batch_size=8, accelerator=comet_device
+        )
+        comet_scores = comet_output["scores"]
+        results_dict["translation"]["full_output"].append(clean_translation)
+        results_dict["translation"]["comet_score"].append(comet_scores[0])
+        results_dict["translation"]["weighted_semantic_density"].append(
+            var_output["translation"]["weighted_semantic_density"]
+        )
+        return results_dict
+
+    def classification_results(self, test_row, _, var_output, results_dict):
+        # ### CLASSIFICATION ###
+        mean_scores = var_output["classification"]["mean_scores"]
+        preds = torch.round(mean_scores).tolist()
+        labels = self.multihot(test_row["labels"])
+        hamming_acc = hamming_loss(y_pred=preds, y_true=labels)
+        zero_one_acc = zero_one_loss(y_pred=preds, y_true=labels)
+        results_dict["classification"]["mean_scores"].append(
+            mean_scores.detach().tolist()
+        )
+        results_dict["classification"]["hamming_accuracy"].append(hamming_acc)
+        results_dict["classification"]["zero_one_accuracy"].append(zero_one_acc)
+        results_dict["classification"]["mean_predicted_entropy"].append(
+            torch.mean(var_output["classification"]["predicted_entropy"]).item()
+        )
+
+        return results_dict
+
+
+def run_inference(
+    dataloader: DataLoader,
+    pipeline: RTCVariationalPipeline | RTCSingleComponentPipeline,
+    results_getter: ResultsGetter,
+):
+    results_dict = {
```

Comment:
why is this here if it gets overwritten?

Reply:
It's only overwritten if there is a single component pipeline, I could instead update it to remove the unused rows?

Reply:
ah ok because get_results updates the results_dict. I hadn't realised that. I don't think it's worth changing here but, In future, I think we should avoid having those sort of side effects and where we do need to have them, make them super clear.

Reply:
Will update the comments to highlight this

Reply:
I've made changes now, instead returning results as a list of dicts, it uses the `celex_id` row of the dataset as a unique identifier.